src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj

```diff
@@ -69,18 +70,36 @@
 
   (mark-environment-expired!
     [this env-name]
-    (let [pool-context (:pool-context (tk-services/service-context this))]
+    (let [environment-class-info (:environment-class-info
+                                  (tk-services/service-context this))
+          pool-context (:pool-context (tk-services/service-context this))]
+      (swap! environment-class-info dissoc env-name)
```

Comment:
I realized after first implementing this that there would be a race condition where if a `mark-*-environment*-expired!` call were to occur while an environment_classes call was right in the middle of parsing info for that environment, the latest environment classes info would be updated based on the result of that environment_classes call - essentially causing this line to be a no-op.  This wouldn't be a problem so long as the environment_classes parsing had started after any code updates on disk had been made, but that's not entirely deterministic without the use of the PE file sync feature.

@cprice404 and I talked about this IRL and we decided, for the sake of getting an initial implementation of this new capability in place, to shore up this race condition as part of a follow-on ticket.  I filed that ticket as SERVER-1130.